### PR TITLE
PR to ease review

### PR DESCRIPTION
**Pull request by zaquestion**

### CODE REVIEW

**Comment by zaquestion** [comment]
Please no dist directory, these are non-diffable refs that only bloat the git tree

Instead create a release and upload the egg:
https://help.github.com/articles/creating-releases/

To create the release (these will work with pypi)
https://github.com/TuneLab/requests-mv-integrations/releases

File: requests_mv_integrations/request_mv_integration.py [hunk 1]
```diff
@@ -0,0 +1,1774 @@
...
+                    request_label=request_label
+                )
+            elif response_content_type.startswith('text/html'):
+                try:
+                    response_content_html_lines = \
+                        requests_response_text_html(
+                            response=response
+                        )
+                except Exception as ex:
+                    raise TuneRequestModuleError(
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: merge this exception with the below one, if the error messages are that important set them with the `if`s and raise at the end

File: requests_mv_integrations/request_mv_integration.py [hunk 2]
```diff
@@ -0,0 +1,1774 @@
...
+        request_extra = {
+            'request_method': request_method,
+            'request_url': request_url,
+            'timeout': timeout,
+            'request_params': safe_dict(request_params),
+            'request_data': request_data_extra,
+            'request_headers': safe_dict(headers)
+        }
+
+        if verbose:
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: We shouldn't be using passing flags around to switch the log level. The 'verbose' flag should be removed from our codebase and log levels set appropriately.

File: requests_mv_integrations/request_mv_integration.py [hunk 3]
```diff
@@ -0,0 +1,1774 @@
...
+            raise TuneRequestError(
+                error_message="Request: Exception: HTTP Error",
+                errors=ex_req_http,
+                exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_HTTP,
+                error_request_curl=self.built_request_curl
+            )
+
+        except requests.exceptions.ConnectionError as ex_req_connect:
+            raise TuneRequestError(
+                error_message="Request: Exception: Connection Error",
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py:425]: Most of these exceptions are the same except for the the message. The error message on the caught exception should be good enough. Suggest consolidating these except chains.

File: requests_mv_integrations/request_mv_integration.py [hunk 4]
```diff
@@ -0,0 +1,1774 @@
...
+        timeout = None
+
+        retry_tries = None
+        retry_delay = None
+        retry_backoff = 0
+        retry_jitter = 0
+        retry_max_delay = None
+
+        if not verify:
+            requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py:287]: This would likely be better placed in __init__.py

File: requests_mv_integrations/request_mv_integration.py [hunk 5]
```diff
@@ -0,0 +1,1774 @@
...
+
+        self.retry_tries = retry_tries
+        self.retry_backoff = retry_backoff
+
+        _tries, _delay, _timeout = retry_tries, retry_delay, timeout
+        while _tries:
+            _attempts += 1
+
+            fkwargs['timeout'] = _timeout
+            request_func = partial(call_func, *args, **kwargs)
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py:572]: This usage of partial should really be cleaned up. Its hard to follow and seeming unneeded. If refactoring the need away is not addressable here I suggest creating the partial in the calling function and only passing the request_func down. The kwarg should be overrideable down here in the function call below

File: requests_mv_integrations/support/tune_request.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    ):
+        response = self.session.request(
+            method=request_method,
+            url=request_url,
+            **kwargs
+        )
+
+        return response
+
+    def request_safe(
```
Comment (zaquestion) [inline requests_mv_integrations/support/tune_request.py]: per our discussion this method should just be removed infavor of the non-safe version

File: requests_mv_integrations/request_mv_integration_download.py [hunk 7]
```diff
@@ -0,0 +1,1235 @@
...
+    RequestMvIntegration
+)
+from logging_mv_integrations import (
+    TuneLoggingFormat
+)
+
+python_check_version(__python_required_version__)
+
+
+class RequestMvIntegrationDownload(RequestMvIntegration):
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration_download.py]: sever the inheritance here and create an instance of RequestMvIntegration

File: requests_mv_integrations/request_mv_integration.py [hunk 8]
```diff
@@ -0,0 +1,1774 @@
...
+                )
+        else:
+            self.logger.debug(
+                "Validate JSON Response: Valid",
+                extra=response_extra
+            )
+
+        return json_response
+
+    @staticmethod
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: for staticmethods these are good candidates to free from the class and "shift to the left"

In cases where we need the logger modify the function signature to accept it.

File: requests_mv_integrations/request_mv_integration.py [hunk 9]
```diff
@@ -0,0 +1,1774 @@
...
+        """
+        tune_request = TuneRequest(
+            retry_tries=self.retry_tries,
+            retry_backoff=self.retry_backoff,
+            retry_codes=self.request_retry_http_status_codes
+        )
+
+        self.session = tune_request.session
+
+        self.logger.debug(
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: I'm not favorable towards these logs that print all of the arguments I just passed to the function. This is something a debugger/tracer is built to handle for you and they are not useful while building an integration. We can also write one off debug statements when we need them, but theres no reason for them to live in our code.

File: requests_mv_integrations/request_mv_integration.py [hunk 10]
```diff
@@ -0,0 +1,1774 @@
...
+            requests.Response
+
+        """
+        tune_request = TuneRequest(
+            retry_tries=self.retry_tries,
+            retry_backoff=self.retry_backoff,
+            retry_codes=self.request_retry_http_status_codes
+        )
+
+        self.session = tune_request.session
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: Would like to see the refs update to just access the session on tune_request.

File: requests_mv_integrations/request_mv_integration.py [hunk 11]
```diff
@@ -0,0 +1,1774 @@
...
+        self,
+        request_method,
+        request_url,
+        request_params=None,
+        request_data=None,
+        request_json=None,
+        request_retry=None,
+        request_retry_excps=None,
+        request_retry_http_status_codes=None,
+        request_retry_func=None,
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py:222]: I would really like this notion of a retry func go away and handled as collector specific code. The HTTPAdapter Retry functionality handles the base 90% of cases.

Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py:222]: @zaquestion 
If we remove **request_retry_func**, then each collector will perform it differently.
Agreed, it is currently used in a limited number of **mv-integrations**, but it is consistent.

File: requests_mv_integrations/request_mv_integration.py (hunk 9 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py]: @zaquestion 
Let us talk about **debugger/tracer** idea tomorrow.

File: requests_mv_integrations/request_mv_integration.py (hunk 1 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py]: @zaquestion 
Will do.
Good catch.

File: requests_mv_integrations/request_mv_integration.py (hunk 2 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py]: @zaquestion 
Will remove.
Good catch.

File: requests_mv_integrations/request_mv_integration.py (hunk 5 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py:572]: @zaquestion 
Will review.

File: requests_mv_integrations/request_mv_integration_download.py (hunk 7 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration_download.py]: Okay.

File: requests_mv_integrations/request_mv_integration.py (hunk 3 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py:425]: @zaquestion 
It is the **exit_code** that is the determining factor how a **mv-integration** exits
In this case:
`exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_CONNECT`

I had removed lots of custom Exceptions, which their only difference was their underlying `exit_code` associated with `TuneIntegrationExitCode`.

File: requests_mv_integrations/request_mv_integration.py (hunk 4 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py:287]: @zaquestion 
If placing this block of code within in `def init(...,verify,...)`, then we are assuming that all requests made by a specific mv-integration would have the same security needs, instead of limited by security needs by request.

I think that security should be disabled as need by request.

File: requests_mv_integrations/request_mv_integration.py (hunk 10 above)
Comment (jefftune) [reply requests_mv_integrations/request_mv_integration.py]: @zaquestion 
Another good catch.
Prior to using `TuneRequest` which only utilizes `Session` object, I was switch between either using `Requests` object or a persistent `Session` object.
Will resolve.

File: requests_mv_integrations/request_mv_integration.py (hunk 4 above)
Comment (zaquestion) [reply requests_mv_integrations/request_mv_integration.py:287]: ah my bad, I thought this something else.

File: requests_mv_integrations/request_mv_integration.py (hunk 3 above)
Comment (zaquestion) [reply requests_mv_integrations/request_mv_integration.py:425]: I realized, I am referring to the blocks that use the same exit code. I will comment on those spots

File: requests_mv_integrations/request_mv_integration.py [hunk 12]
```diff
@@ -0,0 +1,1774 @@
...
+                error_request_curl=self.built_request_curl,
+                exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_CONNECT
+            )
+
+        except BrokenPipeError as ex_broken_pipe:
+            raise TuneRequestError(
+                error_message="Request: Exception: Broken Pipe Error",
+                errors=ex_broken_pipe,
+                error_request_curl=self.built_request_curl,
+                exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_CONNECT
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: also uses `MOD_ERR_REQUEST_CONNECT` as above

File: requests_mv_integrations/request_mv_integration.py [hunk 13]
```diff
@@ -0,0 +1,1774 @@
...
+            )
+
+        except ConnectionError as ex_connect:
+            raise TuneRequestError(
+                error_message="Request: Exception: Connection Error",
+                errors=ex_connect,
+                error_request_curl=self.built_request_curl,
+                exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_CONNECT
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: also uses `MOD_ERR_REQUEST_CONNECT` as above

File: requests_mv_integrations/request_mv_integration.py [hunk 14]
```diff
@@ -0,0 +1,1774 @@
...
+            )
+
+        except requests.packages.urllib3.exceptions.ProtocolError as ex_req_urllib3_protocol:
+            raise TuneRequestError(
+                error_message="Request: Exception: Protocol Error",
+                errors=ex_req_urllib3_protocol,
+                error_request_curl=self.built_request_curl,
+                exit_code=TuneIntegrationExitCode.MOD_ERR_REQUEST_CONNECT
```
Comment (zaquestion) [inline requests_mv_integrations/request_mv_integration.py]: also uses `MOD_ERR_REQUEST_CONNECT` as above
